cloud-controller-manager/do/firewall_controller_diff.go

```diff
@@ -0,0 +1,127 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package do
+
+import (
+	"strings"
+
+	"github.com/digitalocean/godo"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+// comparableFirewall is a struct holding just the fields we care about during
+// comparison.
+type comparableFirewall struct {
+	Name          string
+	InboundRules  []godo.InboundRule
+	OutboundRules []godo.OutboundRule
+	Tags          []string
+}
+
+func firewallRequestEqual(fw *godo.Firewall, fr *godo.FirewallRequest) (bool, string) {
+	cf1 := compFirewallFromFirewall(fw)
+	cf2 := compFirewallFromFirewallRequest(fr)
+
+	return compFirewallsEqual(cf1, cf2)
+}
+
+func firewallsEqual(fw1, fw2 *godo.Firewall) (bool, string) {
+	cf1 := compFirewallFromFirewall(fw1)
+	cf2 := compFirewallFromFirewall(fw2)
+
+	return compFirewallsEqual(cf1, cf2)
+}
+
+func compFirewallsEqual(cf1, cf2 *comparableFirewall) (bool, string) {
+	// Equality when both variables are empty.
+	if cf1 == nil && cf2 == nil {
+		return true, ""
+	}
+
+	// Non-equality when exactly one of two variables is empty.
+	if cf1 == nil || cf2 == nil {
+		return false, ""
+	}
+
+	// Define custom sorters to guard against non-deterministic rule sort orders.
+	sorterInboundRules := cmpopts.SortSlices(func(r1, r2 godo.InboundRule) bool {
+		return printInboundRule(r1) < printInboundRule(r2)
+	})
+	sorterOutboundRules := cmpopts.SortSlices(func(r1, r2 godo.OutboundRule) bool {
+		return printOutboundRule(r1) < printOutboundRule(r2)
+	})
+
+	// A PortRange indicates "all ports" when the value is set to "all" on requests,
+	// and the corresponding response returns "0" then. Map them all to a single
+	// value for comparing.
+	portRangeMapper := cmp.FilterPath(func(p cmp.Path) bool {
+		switch p.String() {
+		case "InboundRules.PortRange", "OutboundRules.PortRange":
+			return true
+		}
+		return false
+	}, cmp.Comparer(func(pr1, pr2 string) bool {
+		sanitizePortRange := func(pr string) string {
+			if pr == "" || pr == "0" || pr == "all" {
+				pr = "0"
+			}
+			return pr
+		}
+
+		return sanitizePortRange(pr1) == sanitizePortRange(pr2)
+	}))
+
+	// Ignore all fields on {In,Out}boundRules.{Sources,Destinations} other than
+	// "Addresses".
+	ruleSourceDestFilter := cmp.FilterPath(func(p cmp.Path) bool {
+		if strings.HasPrefix(p.String(), "InboundRules.Sources") || strings.HasPrefix(p.String(), "OutboundRules.Destinations") {
+			return p.Last().String() != ".Addresses"
+		}
+
+		return false
+	}, cmp.Ignore())
+
+	diff := cmp.Diff(cf1, cf2, sorterInboundRules, sorterOutboundRules, portRangeMapper, ruleSourceDestFilter, cmpopts.EquateEmpty())
+	return diff == "", diff
+}
+
+func compFirewallFromFirewall(fw *godo.Firewall) *comparableFirewall {
```

Comment:
This function and the next one deserve a comment of why this auxiliary struct is being built. Alternatively I'd move them up to the struct which has an explanation for that.
An even cleaner solution would be to take those two functions and convert them into transformers for the `cmp.Diff` call.

Reply:
Not 100% sure about what would all need to change to make the two functions produce transformers. I'll move them further up for now, but curious to hear details about the transformer suggestion (either here or via DM).